assets/src/components/mapPage/mapDisplay.tsx

```diff
@@ -0,0 +1,392 @@
+              <RouteVehicles
+                selectedVehicleRoute={selectedVehicleOrGhost.routeId}
+                selectedVehicleId={selectedVehicleOrGhost.id}
+                onPrimaryVehicleSelect={selectVehicle}
+              />
+            ))}
+        </>
+      )}
+
+      <InterruptibleFollower
+        onUpdate={onFollowerUpdate}
+        positions={position}
+        {...followerState}
+      />
+    </>
+  )
+}
+
+const SelectionDataLayers = ({
+  liveSelectedEntity,
+  showSelectionCard,
+  deleteSelection,
+  setSelection,
+  setStateClasses,
+}: {
+  liveSelectedEntity: LiveSelectedEntity | null
+  showSelectionCard: boolean
+  deleteSelection: () => void
+  setSelection: (selectedEntity: SelectedEntity | null) => void
+  setStateClasses: (classes: string | undefined) => void
+}) => {
+  switch (liveSelectedEntity?.type) {
+    case SelectedEntityType.Vehicle:
+      return (
+        <SelectedVehicleDataLayers
+          vehicleOrGhost={liveSelectedEntity.vehicleOrGhost}
+          showSelectionCard={showSelectionCard}
+          deleteSelection={deleteSelection}
+          selectVehicle={(vehicleOrGhost: VehicleOrGhost) =>
+            setSelection({
+              type: SelectedEntityType.Vehicle,
+              vehicleId: vehicleOrGhost.id,
+            })
+          }
+          setStateClasses={setStateClasses}
+        />
+      )
+    // TODO: handle SelectedEntityType.ROUTE
+    default:
+      return <MapElementsNoSelection setStateClasses={setStateClasses} />
+  }
+}
+
+const shouldShowShape = (liveSelectedEntity: LiveSelectedEntity | null) => {
+  switch (liveSelectedEntity?.type) {
+    case SelectedEntityType.RoutePattern:
+      return true
+    case SelectedEntityType.Vehicle:
+      return (
+        liveSelectedEntity.vehicleOrGhost &&
+        isVehicle(liveSelectedEntity.vehicleOrGhost) &&
+        !liveSelectedEntity.vehicleOrGhost.isShuttle
+      )
+    default:
+      return false
+  }
+}
+
+const MapDisplay = ({
+  selectedEntity,
+  setSelection,
+  showSelectionCard,
+}: {
+  selectedEntity: SelectedEntity | null
+  setSelection: (selectedEntity: SelectedEntity | null) => void
+  showSelectionCard: boolean
+}) => {
+  const deleteSelection = useCallback(() => {
+    setSelection && setSelection(null)
+  }, [setSelection])
+
+  const stations = useStations()
+
+  const liveSelectedEntity: LiveSelectedEntity | null =
+    useLiveSelectedEntity(selectedEntity)
+
+  const routePatternIdentifier =
+    routePatternIdentifierForSelection(liveSelectedEntity)
+
+  const allPatternsForRoute: ByRoutePatternId<RoutePattern> | null =
+    usePatternsByIdForRoute(routePatternIdentifier?.routeId || null)
+
+  const [stateClasses, setStateClasses] = useState<string | undefined>(
+    undefined
+  )
+
+  const routePatternShape =
+    allPatternsForRoute &&
+    routePatternIdentifier &&
+    shouldShowShape(liveSelectedEntity)
+      ? allPatternsForRoute[routePatternIdentifier.routePatternId]?.shape
+      : null
```

Comment:
I was **really** hoping to do away with the `?` in `allPatternsForRoute[routePatternIdentifier.routePatternId]?.shape` here. I want it to be guaranteed that allPatternsForRoute contains the routePatternId of the selected vehicle/route. 

The only time it doesn't is when switching between vehicles: the routePatternIdentifier may be of the new vehicle, but allPatternsForRoute may lag behind with the patterns for the old vehicle. I was considering making this more transactional, but I'm thinking that may be out of scope based on our sync discussion on handling intermediate loading states.